anonlink/similarities.py

```diff
@@ -0,0 +1,100 @@
+import collections
+import functools
+import heapq
+import itertools
+import operator
+from typing import Counter, DefaultDict, Dict, List, Optional, Sequence, Tuple
+
+import numpy as np
+
+from .typechecking import Record
+
+
+def _hamming_sim_f(clk1: Record, clk2: Record) -> float:
+    assert len(clk1) == len(clk2)
+    assert len(clk1) != 0
+    return sum(map(operator.eq, clk1, clk2)) / len(clk1)
+
+
+def _hamming_similarity_k(
+    datasets: Sequence[Sequence[Record]],
+    threshold: float,
+    k: int
+) -> Tuple[Sequence[Sequence[int]], Sequence[float]]:
+    # Dictionaries support sparsity, but these could also be list...
+    candidates = tuple(collections.defaultdict(list) for _ in datasets)  # type: Tuple[DefaultDict[int, List], ...]
+
+    for records in itertools.product(enumerate(datasets[0]),
+                                     enumerate(datasets[1])):
+        (i0, clk0), (i1, clk1) = records
+        sim = _hamming_sim_f(clk0, clk1)
+
+        if sim >= threshold:
+            c = sim, (i0, i1)
+            candidates[0][i0].append(c)
+            candidates[1][i1].append(c)
+
+    # Take the best k candidates for each record and count them
+    pair_counter = collections.Counter()  # type: Counter[Tuple[float, Tuple[int, int]]]
+    for dset_cands in candidates:
+        for unfiltered_cands in dset_cands.values():
+            pair_counter.update(heapq.nlargest(k, unfiltered_cands))
+
+    result_pairs = tuple(filter(lambda pair: pair_counter[pair] == 2,
+                                pair_counter))
+    # Make into arrays
+    if result_pairs:
+        sims, indices = zip(*result_pairs)
+        sims_arr = np.array(sims, dtype=float)
+        indices_arr = np.array(indices, dtype=int, order='F').T
+    else:
+        sims_arr = np.empty((0,), dtype=float)
+        indices_arr = np.empty((2,0), dtype=int)
+
+    return indices_arr, sims_arr
+
+
+def _hamming_similarity_no_k(
```

Comment:
Why not just handle the 'no k' case in the other function with an if statement to skip the `pair_counter` business?

Reply:
That’s what I had done originally, but I found that it took multiple `if` statements, and having them interspersed through the code made the function look messy. Plus, there are extra optimisations to be made if `k is None`.